blog/2020/05/rr.md

```diff
@@ -0,0 +1,452 @@
+
+$$ \mathcal{S}_{i+1} = \begin{cases}
+    \mathcal{A}_i(S_{i}) & \text{if an asynchronous event occured at time $i$} \\
+    \mathcal{I}_{\text{ip}(\mathcal{S}_i)} & \text{otherwise}
+\end{cases} $$
+
+Did that help? No? oh. Well, it made me feel better about my physics degree and
+also I've been told it's not research until it has at least one equation in it,
+so there you go. Also, people worked really hard to get the equation rendering working,
+on this blog, so I better use it. Anyway, where were we?
+
+Ah yes, `rr` works by treating the kernel/userspace boundary as the determinism boundary.
+This has a lot of advantages. For one, the kernel isolates a lot of state. If a
+part of the disk state (e.g. a file) wasn't requested through the kernel, you
+can be guaranteed that it didn't affect the process state (assuming the kernel
+works correctly of course). It abstracts over hardware details and provides
+a uniform interface to resources like the network. It also isolates processes
+from another, so if there is one process of interest (e.g. julia) only the
+activity relevant to it has to be recovered.
+
+Of course, there are some drawbacks as well. In order to function `rr` has to
+have an extremely precise model of the operation of the kernel and the ways
+in which it interacts with userspace. On Linux this interface is supposedly
+stable, but few applications put as stringent a constraint on that promise as
+`rr` does (e.g. even a single bit difference can be problematic).
+
+The details here are quite involved, and I recommend reading Robert's
+[technical report](https://arxiv.org/abs/1610.02144) for a deeper overview of
+some of `rr`'s design points (though even that only scratches the surface).
+
+One final thought for this section is the origin of the term "time-traveling"
+debugger. It stems from the mode of analysis that a system like this enables.
+In a traditional debugger, it is possible to execute one instruction at a time,
+and step from memory state to memory state in the forwards direction. Systems
+like `rr` allow the opposite during replay: Step backwards through the state of
+the system. Under the hood this is accomplished by playing forward from the
+beginning (or more likely some intermediate checkpoint created for performance
+reasons), until the previous state is reached. However, to the end user, the
+illusion of going backwards in time is presented and an extremely useful mental
+model for debugging.
+
+# Performance considerations
+
+While the previous section described the operating principle of `rr`, it does
+not do justice to the reason why `rr` works so well. That reason is simple:
+performance. Overhead for recording of single threaded processes
+is generally below 2x, most often between 2% and 50% (lower for purely
+numerical calculations, higher for workloads that interact with the OS).
+Recording multi-threaded processes is tougher. By default `rr` serializes
```

Comment:
```suggestion
Recording multiple threads or processes is harder. Instead, `rr` serializes the execution of each thread. This incurs overhead linear in the number of threads. But, in generality, this is required for reproducibility. It is thus a good idea to try and reproduce any issues at low core count. There is interesting academic research in other techniques, but nothing is close to production-ready yet.
```
(and kill the rest of this paragraph, or at least save it for your JuliaCon presentation)

Reply:
Fair. Applied in spirit, but with slight rewordings for flow.